src/rgw/cls_fifo_legacy.cc

```diff
@@ -1153,6 +1184,20 @@ void FIFO::trim_part(const DoutPrefixProvider *dpp, int64_t part_num, uint64_t o
   ceph_assert(r >= 0);
 }
 
+bool FIFO::_need_new_head(const DoutPrefixProvider* dpp, std::unique_lock<std::mutex>& l) {
+  ldpp_dout(dpp, 20)
+    << __PRETTY_FUNCTION__ << ":" << __LINE__
+    << " entering" << dendl;
+  assert(l.owns_lock());
+  if (preparing) {
+    ldpp_dout(dpp, 20)
+      << __PRETTY_FUNCTION__ << ":" << __LINE__
+      << " someone else is making a new head, waiting" << dendl;
+    cond.wait(l, [this] { return !preparing; });
```

Comment:
Maybe. I think I'll rework things so if we do need a new head we'll hold the lock straight into `_prepare_new_head`, since I also don't want a condition where we legitimately do need a new head but don't make one.